lib/recursive-methods.rb

```diff
@@ -1,49 +1,99 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n!)
+# Space complexity: O(n!)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    if n == 1
+        return 1
+    elsif n == 0
+        return 1
+    else
+        return n * factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    new_string = ""
+    s.length.times do |i|
+        new_string += s[-1 - i]
+    end
+    return new_string
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
```

Comment:
Since you are making a new string with each recursive call this is O(n^2) space complexity.

Also **not** in place